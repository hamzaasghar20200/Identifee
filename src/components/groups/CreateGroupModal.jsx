import React, { useEffect, useState } from 'react';
import {
  CardBody,
  CardFooter,
  Col,
  Form,
  FormGroup,
  Input,
  Label,
} from 'reactstrap';

import RightPanelModal from '../modal/RightPanelModal';
import stringConstants from '../../utils/stringConstants.json';
import { RIGHT_PANEL_WIDTH, isAlphanumeric } from '../../utils/Utils';
import { DropdownTreeView } from '../prospecting/v2/common/DropdownTreeView';
import { useForm } from 'react-hook-form';
import InputValidationAdvance from '../commons/InputValidationAdvance';
import ButtonIcon from '../commons/ButtonIcon';
import ControllerValidation from '../commons/ControllerValidation';

const errorAlphanumeric = stringConstants.settings.users.filters.alphanumeric;

const CreateGroupModal = ({
  setErrorMessage,
  showModal,
  setShowModal,
  createGroup,
  data = [],
  siblingAccess,
  setIsDropdownId,
  isDropdownId,
  isAddSingleRole = '',
  perantId,
  isLoading,
  setSblingAccess,
}) => {
  const groupObj = {
    group_name: '',
    parent_id: '',
    description: '',
  };
  const [groupForm, setGroupForm] = useState(groupObj);
  const {
    register,
    handleSubmit,
    reset,
    setValue,
    control,
    getFieldState,
    formState: { errors },
  } = useForm({
    defaultValues: groupForm,
  });
  const constants = stringConstants.settings.groups;

  const alphanumericError = (input) => {
    const msgError = errorAlphanumeric.error;
    if (input === 'search') {
      setErrorMessage(msgError);
      setTimeout(() => setErrorMessage(''), 3500);
    }
  };
  const onInputChange = (e) => {
    const { value, name } = e.target || {};
    isAlphanumeric(value)
      ? setGroupForm({ ...groupForm, [name]: value })
      : alphanumericError(`search`);
    setValue(name, value);
  };

  // Handler of submit
  const onHandleSubmit = () => {
    createGroup(groupForm);
    setIsDropdownId('');
    reset(
      setGroupForm({
        group_name: '',
        parent_id: null,
        description: '',
      })
    );
    setSblingAccess(false);
    setShowModal(false);
  };

  const closeModal = () => {
    setShowModal(false);
    reset(
      setGroupForm({
        group_name: '',
        parent_id: null,
        description: '',
      })
    );
    setIsDropdownId('');
  };
  useEffect(() => {
    if (perantId) {
      setGroupForm({
        ...groupForm,
        parent_id: perantId,
      });
      setValue('parent_id', perantId);
    }
  }, [perantId]);
  console.log(isAddSingleRole);
  return (
    <RightPanelModal
      Title={constants.create.addGroupModalTitle}
      onHandleCloseModal={() => closeModal()}
      showModal={showModal}
      buttonLabel={constants.create.roleAddGroup}
      handleSubmit={() => handleSubmit()}
      allowCloseOutside={false}
      setShowModal={setShowModal}
      showOverlay={true}
      containerBgColor={'pb-0'}
      containerWidth={RIGHT_PANEL_WIDTH}
      containerPosition={'position-fixed'}
      headerBgColor="bg-gray-5"
    >
      <CardBody>
        <Form onSubmit={handleSubmit(onHandleSubmit)}>
          <InputValidationAdvance
            fieldType="text"
            type="input"
            label="Role Name"
            name="group_name"
            id="group_name"
            classNames={`border-left-4 border-left-danger`}
            className="mt-2 mb-2"
            value={groupForm}
            errors={errors}
            validationConfig={{
              required: true,
              inline: false,
              onChange: (e) => {
                onInputChange(e);
              },
            }}
            register={register}
            placeholder="Role Name"
          />

          <FormGroup row className="align-items-center">
            <Label md={3} className="mt-0 text-right">
              Reports to
            </Label>
            <Col md={9} className="pl-0">
              <ControllerValidation
                name={'parent_id'}
                errors={errors}
                form={groupForm}
                errorDisplay="mb-0"
                control={control}
                validationConfig={{
                  required: 'Role Id is Required',
                }}
                renderer={({ field }) => (
                  <DropdownTreeView
                    data={data}
                    validationConfig={{
                      required: true,
                      inline: false,
                    }}
                    fieldState={getFieldState('parent_id')}
                    setIsDropdownId={setIsDropdownId}
                    isDropdownId={isDropdownId}
                    editRoleData={isAddSingleRole}
                  />
                )}
              />
            </Col>
          </FormGroup>
          <FormGroup row className="mx-0">
            <Label
              md={9}
              htmlFor="has_sibling_access"
              className="form-label ml-auto"
            >
              <Input
                type="checkbox"
                name="has_sibling_access"
                id="has_sibling_access"
                value={siblingAccess}
                checked={siblingAccess}
                onChange={() => setSblingAccess(!siblingAccess)}
              />{' '}
              <span className="font-weight-normal">
                {' '}
                Let users in this role see each other&apos;s data
              </span>
            </Label>
          </FormGroup>
          <InputValidationAdvance
            fieldType="text"
            type="textarea"
            label="Description"
            name="description"
            id="description"
            rows={3}
            classNames={`border-left-4 border-left-danger`}
            className="mt-2 mb-2"
            value={groupForm}
            errors={errors}
            validationConfig={{
              required: true,
              inline: false,
              onChange: (e) => {
                onInputChange(e);
              },
            }}
            register={register}
            placeholder="A few words about this role"
          />
        </Form>
      </CardBody>
      <CardFooter className="text-right">
        <ButtonIcon
          label="Cancel"
          type="button"
          color="white"
          classnames="btn-white mx-1 btn-sm"
          onclick={showModal ? closeModal : ''}
        />
        <ButtonIcon
          classnames="btn-sm"
          type="button"
          onclick={handleSubmit(onHandleSubmit)}
          label={'Save'}
          color={`primary`}
          loading={isLoading}
        />
      </CardFooter>
    </RightPanelModal>
  );
};

export default CreateGroupModal;
