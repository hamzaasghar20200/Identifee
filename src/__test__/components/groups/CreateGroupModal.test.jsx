import { fireEvent, render } from '@testing-library/react';
import { waitFor } from '@testing-library/dom';

import CreateGroupModal from '../../../components/groups/CreateGroupModal';
import stringConstants from '../../../utils/stringConstants.json';

describe('Test CreateGroupModal', () => {
  const constants = stringConstants.settings.groups;
  const props = {
    showModal: true,
    data: [
      { id: 1, name: 'Group 1' },
      { id: 2, name: 'Group 2' },
      { id: 3, name: 'Group 3' },
    ],
  };

  beforeEach(() => {
    jest.resetModules();
    jest.restoreAllMocks();
    jest.resetAllMocks();
    jest.clearAllMocks();
  });

  it('Snapshot', () => {
    const component = render(<CreateGroupModal {...props} />);
    expect(component.container).toBeVisible();
    expect(component.baseElement).toMatchSnapshot();
  });

  it('Snapshot with dropdown menu open', async () => {
    const component = render(<CreateGroupModal {...props} />);

    const buttonDropdown = component.getByText(
      constants.create.dropTextParentGroup
    );
    fireEvent.click(buttonDropdown);

    await waitFor(() => {
      expect(buttonDropdown).toBeVisible();
      expect(component.baseElement).toMatchSnapshot();
    });
  });

  it('Should call prop setShowModal when try close modal', async () => {
    const mockCloseHandler = jest.fn();
    const component = render(
      <CreateGroupModal {...props} setShowModal={mockCloseHandler} />
    );

    const iconClose = document.getElementsByClassName('close')[0];
    fireEvent.click(iconClose);

    await waitFor(() => {
      expect(iconClose).toBeVisible();
      expect(mockCloseHandler).toHaveBeenCalledTimes(1);
    });

    const buttonClose = component.getByText('Cancel');
    fireEvent.click(buttonClose);

    await waitFor(() => {
      expect(buttonClose).toBeVisible();
      expect(mockCloseHandler).toHaveBeenCalledTimes(2);
    });
  });

  it('Should call prop onChangeDrop when writes in the input from dropdown', async () => {
    const mockOnChangeDrop = jest.fn();
    const component = render(
      <CreateGroupModal {...props} onChangeDrop={mockOnChangeDrop} />
    );

    const buttonDropdown = component.getByText(
      constants.create.dropTextParentGroup
    );
    fireEvent.click(buttonDropdown);

    const dropdownMenu = document.getElementsByClassName('dropdown-menu')[0];
    const inputMenu = dropdownMenu.getElementsByClassName('form-control')[0];

    fireEvent.change(inputMenu, { target: { value: 'Group' } });

    await waitFor(() => {
      expect(dropdownMenu).toBeVisible();
      expect(inputMenu).toBeVisible();
      expect(mockOnChangeDrop).toHaveBeenCalledTimes(1);
    });
  });

  it('Should allow select an item from dropdown', async () => {
    const component = render(<CreateGroupModal {...props} />);

    const buttonDropdown = component.getByText(
      constants.create.dropTextParentGroup
    );
    fireEvent.click(buttonDropdown);

    const firstItem = component.getByText(props.data[0].name);
    fireEvent.click(firstItem);

    await waitFor(() => {
      expect(buttonDropdown).toBeVisible();
      expect(buttonDropdown).toHaveTextContent(props.data[0].name);
    });
  });

  it('Should call prop createGroup when it has the items selected', async () => {
    const mockCreateGroup = jest.fn();
    const mockCloseHandler = jest.fn();
    const component = render(
      <CreateGroupModal
        {...props}
        createGroup={mockCreateGroup}
        setShowModal={mockCloseHandler}
      />
    );

    const buttonDropdown = component.getByText(
      constants.create.dropTextParentGroup
    );
    fireEvent.click(buttonDropdown);

    const firstItem = component.getByText(props.data[0].name);
    fireEvent.click(firstItem);

    const inputGroup = component.getByTestId('group_name');
    fireEvent.change(inputGroup, { target: { value: 'New group' } });

    await waitFor(() => {
      expect(buttonDropdown).toHaveTextContent(props.data[0].name);
      expect(inputGroup.value).toContain('New group');
    });

    const buttonCreate = component.getByText(constants.create.btnAddGroup);
    fireEvent.click(buttonCreate);

    await waitFor(() => {
      expect(mockCreateGroup).toHaveBeenCalledTimes(1);
      expect(mockCloseHandler).toHaveBeenCalledTimes(1);
    });
  });
});
