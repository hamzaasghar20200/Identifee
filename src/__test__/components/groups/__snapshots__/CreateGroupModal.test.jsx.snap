// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test CreateGroupModal Snapshot 1`] = `
<body
  class="modal-open"
>
  <div />
  <div
    style="position: relative; z-index: 1050; display: block;"
    tabindex="-1"
  >
    <div
      class=""
    >
      <div
        class="modal fade"
        role="dialog"
        style="display: block;"
        tabindex="-1"
      >
        <div
          class="modal-dialog modal-dialog-centered"
          role="document"
        >
          <div
            class="modal-content"
          >
            <div
              class="modal-header"
            >
              <h2
                class="modal-title"
              >
                Add New Group
              </h2>
              <button
                aria-label="Close"
                class="close"
                type="button"
              >
                <span
                  aria-hidden="true"
                >
                  ×
                </span>
              </button>
            </div>
            <div
              class="modal-body"
            >
              <span
                class="font-size-sm"
              >
                Parent Group
              </span>
              <input
                class="mt-2 mb-2 form-control"
                data-testid="group_name"
                id="group_name"
                name="group_name"
                placeholder="Field name"
                type="text"
                value=""
              />
              <span
                class="font-size-sm"
              >
                Parent Group
              </span>
              <div
                class="mt-2 dropdown"
                id="dropdown-assigned_parent"
              >
                <button
                  aria-expanded="false"
                  aria-haspopup="true"
                  class="w-100 dropdown-search false dropdown-toggle btn btn-outline-link"
                  type="button"
                >
                  Select Parent Group
                </button>
              </div>
            </div>
            <div
              class="modal-footer"
            >
              <button
                class="btn btn-white"
              >
                Cancel
              </button>
              <button
                class="btn btn-primary"
                disabled=""
                type="button"
              >
                <span>
                  Create
                </span>
              </button>
            </div>
          </div>
        </div>
      </div>
      <div
        class="modal-backdrop fade"
      />
    </div>
  </div>
</body>
`;

exports[`Test CreateGroupModal Snapshot with dropdown menu open 1`] = `
<body
  class="modal-open"
>
  <div />
  <div
    style="position: relative; z-index: 1050; display: block;"
    tabindex="-1"
  >
    <div
      class=""
    >
      <div
        class="modal fade"
        role="dialog"
        style="display: block;"
        tabindex="-1"
      >
        <div
          class="modal-dialog modal-dialog-centered"
          role="document"
        >
          <div
            class="modal-content"
          >
            <div
              class="modal-header"
            >
              <h2
                class="modal-title"
              >
                Add New Group
              </h2>
              <button
                aria-label="Close"
                class="close"
                type="button"
              >
                <span
                  aria-hidden="true"
                >
                  ×
                </span>
              </button>
            </div>
            <div
              class="modal-body"
            >
              <span
                class="font-size-sm"
              >
                Parent Group
              </span>
              <input
                class="mt-2 mb-2 form-control"
                data-testid="group_name"
                id="group_name"
                name="group_name"
                placeholder="Field name"
                type="text"
                value=""
              />
              <span
                class="font-size-sm"
              >
                Parent Group
              </span>
              <div
                class="mt-2 show dropdown"
                id="dropdown-assigned_parent"
              >
                <button
                  aria-expanded="true"
                  aria-haspopup="true"
                  class="w-100 dropdown-search false dropdown-toggle btn btn-outline-link"
                  type="button"
                >
                  Select Parent Group
                </button>
                <div
                  aria-labelledby=""
                  class="basic-animation border border-1 overflow-auto w-100 dropdown-menu show"
                  style="position: absolute; top: 0px; left: 0px; opacity: 0; pointer-events: none; margin: 0px;"
                  x-placement="bottom-start"
                >
                  <div
                    class="col-12"
                  >
                    <input
                      class="form-control"
                      placeholder="Select Parent Group"
                      type="text"
                    />
                  </div>
                  <div
                    class="px-0 dropdown-results pt-2 pl-3 px-3 col-12"
                  >
                    <a
                      class="p-2 item-btn rounded w-100 basic-animation tr-hover cursor-pointer dropdown-item"
                      data-testid="item-1"
                      href="#"
                      id="item-1"
                      role="button"
                    >
                      <div
                        class="user-avatar-select w-100"
                      >
                        <div
                          class="item-user text-capitalize mx-0 w-100 col"
                        >
                          <h5
                            class="mb-0 w-100"
                          >
                            Group 1
                          </h5>
                        </div>
                      </div>
                    </a>
                    <a
                      class="p-2 item-btn rounded w-100 basic-animation tr-hover cursor-pointer dropdown-item"
                      data-testid="item-2"
                      href="#"
                      id="item-2"
                      role="button"
                    >
                      <div
                        class="user-avatar-select w-100"
                      >
                        <div
                          class="item-user text-capitalize mx-0 w-100 col"
                        >
                          <h5
                            class="mb-0 w-100"
                          >
                            Group 2
                          </h5>
                        </div>
                      </div>
                    </a>
                    <a
                      class="p-2 item-btn rounded w-100 basic-animation tr-hover cursor-pointer dropdown-item"
                      data-testid="item-3"
                      href="#"
                      id="item-3"
                      role="button"
                    >
                      <div
                        class="user-avatar-select w-100"
                      >
                        <div
                          class="item-user text-capitalize mx-0 w-100 col"
                        >
                          <h5
                            class="mb-0 w-100"
                          >
                            Group 3
                          </h5>
                        </div>
                      </div>
                    </a>
                  </div>
                </div>
              </div>
            </div>
            <div
              class="modal-footer"
            >
              <button
                class="btn btn-white"
              >
                Cancel
              </button>
              <button
                class="btn btn-primary"
                disabled=""
                type="button"
              >
                <span>
                  Create
                </span>
              </button>
            </div>
          </div>
        </div>
      </div>
      <div
        class="modal-backdrop fade"
      />
    </div>
  </div>
</body>
`;
